nearlib/wallet-account.js

```diff
@@ -0,0 +1,150 @@
+
+        this._initHtmlElements();
+        this._signatureRequests = {};
+        this._authData = JSON.parse(window.localStorage.getItem(LOCAL_STORAGE_KEY) || '{}');
+
+        if (!this.isSignedIn()) {
+            this._tryInitFromUrl();
+        }
+    }
+
+    isSignedIn() {
+        // Later it should call wallet to check auth token is still up to date.
+        return !!this._authData.accountId;
+    }
+
+    getAccountId() {
+        return this._authData.accountId || '';
+    }
+
+    requestSignIn(contract_id, title, success_url, failure_url) {
+        const currentUrl = new URL(window.location.href);
+        let newUrl = new URL(this._walletBaseUrl + LOGIN_WALLET_URL_SUFFIX);
+        newUrl.searchParams.set('title', title);
+        newUrl.searchParams.set('contract_id', contract_id);
+        newUrl.searchParams.set('success_url', success_url || currentUrl.href);
+        newUrl.searchParams.set('failure_url', failure_url || currentUrl.href);
+        newUrl.searchParams.set('app_url', currentUrl.origin);
+        window.location.replace(newUrl.toString());
+    }
+
+    signOut() {
+        this._authData = {};
+        window.localStorage.removeItem(LOCAL_STORAGE_KEY);
+    }
+
+    _tryInitFromUrl() {
+        let currentUrl = new URL(window.location.href);
+        let authToken = currentUrl.searchParams.get('auth_token') || '';
+        let accountId =currentUrl.searchParams.get('account_id') || '';
+        if (!!authToken && !!accountId) {
+            this._authData = {
+                authToken,
+                accountId,
+            };
+            window.localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(this._authData));
+        }
+    }
+
+    _initHtmlElements() {
+        // Wallet iframe
+        const iframe = document.createElement('iframe');
+        iframe.style = 'display: none;';
+        iframe.src = this._walletBaseUrl + EMBED_WALLET_URL_SUFFIX;
+        document.body.appendChild(iframe);
+        this._walletWindow = iframe.contentWindow;
+
+        // Message Event
+        window.addEventListener('message', this.receiveMessage.bind(this), false);
+    }
+
+    receiveMessage(event) {
+        if (event.origin != this._walletBaseUrl) {
+            // Only processing wallet messages.
+            return;
+        }
+        let data;
+        try {
+            data = JSON.parse(event.data);
+        } catch (e) {
+            console.error('Can\'t parse the result', event.data, e);
+            return;
+        }
+        const request_id = data.request_id || '';
+        if (!(request_id in this._signatureRequests)) {
+            console.error('Request ID' + request_id + ' was not found');
+            return;
+        }
+        let signatureRequest = this._signatureRequests[request_id];
+        delete this._signatureRequests[request_id];
+
+        if (data.success) {
+            signatureRequest.resolve(data.result);
+        } else {
+            signatureRequest.reject(data.error);
```

Comment:
need to call `.reject` on timeout as well, to avoid caller hanging forever waiting for promise

Reply:
Added a todo